Fix MeshNormals and WireframeFilter with empty MeshData

brisvag opened the pull request:
When MeshNormal is provided an empty `MeshData` object, all sorts of issues arise, crashing the program. This is especially problematic if one wants to automatically update a `MeshNormals` based on a `Mesh`, which is totally fine with empty data.

I tried fixing this by making the methods in `MeshData` used by `MeshNormals` (such as `get_vertex_normals`) more resilient, but I encountered all sorts of issues. The most straight forward way to solve it ended up being this.

As an added bonus, this PR allows to use `set_data` to update properties of the `MeshNormals` object which are not accounted for by the `LineVisual` (for example, `length`).

~It's missing some tests, but it should be funcional.~

EDT: I added a similar check for `WireframeFilter`, which was failing when `_update_data` was getting triggered (for example by setting `enabled = True`) if `MeshData` was empty. Added a test for this which fails on main.

djhoese commented on vispy/visuals/mesh_normals.py:
```diff
@@ -76,14 +76,59 @@ class MeshNormalsVisual(LineVisual):
 
         if primitive == 'face':
-            normals = meshdata.get_face_normals()
+            if meshdata._faces is None:
```
Could you change this to `meshdata.get_faces()` and similarly the vertices check below? Avoid using the private API. Or would it be good to have a `__nonzero__` or `is_empty` method added to MeshData and then you could check that before these face/vertex if statements.

Additionally, could you add an `else:` (or an extra check above) for the case where `primitive` is not face or vertex. As is right now `normals` could not exist by the time it gets to `if normals is None` below.

brisvag replied:
That can be done for faces, but `get_vertices` computes some stuff and that was at the root of some of the issues.

brisvag replied:
The check for `primitive` already exists above!

djhoese replied:
What do you think about the is_empty-like checks on the MeshData itself? Something to hide the private property checks.

brisvag replied:
Sounds reasonable!

brisvag commented:
Added an `is_empty` check; as far as I can tell, not having faces is enough for not being able to calculate normals (including vertex normals).

I also added a test, which I confirm fails on main.

brisvag commented:
Is that failure as unrelated as it seems?

djhoese commented:
Unrelated? Yes.

A problem? Yes. I restarted the jobs and hope that it is some weird hiccup but otherwise this may mean we have to rethink osmesa testing. CC @larsoner who I think originally got the osmesa tests running (or maybe @kmuehlbauer )?

djhoese commented:
For the record:

```
 Cleaning build, pyc, ctags, and cache
python make test osmesa
WARNING: VISPY_GL_LIB is ignored when using OSMesa. Use OSMESA_LIBRARY instead.
Traceback (most recent call last):
/home/runner/micromamba/envs/vispy-tests/lib/libOSMesa32.so: cannot open shared object file: No such file or directory
  File "make/make.py", line 105, in test
    test(args[0], ' '.join(args[1:]), coverage=True)
  File "/home/runner/work/vispy/vispy/vispy/util/wrappers.py", line 153, in test
    return _test(*args, **kwargs)
  File "/home/runner/work/vispy/vispy/vispy/testing/_runners.py", line 384, in test
    from ..app.backends import BACKEND_NAMES as backend_names
  File "/home/runner/work/vispy/vispy/vispy/app/__init__.py", line 13, in <module>
    from .canvas import Canvas, MouseEvent, KeyEvent  # noqa
  File "/home/runner/work/vispy/vispy/vispy/app/canvas.py", line 16, in <module>
    from ..gloo.context import (GLContext, set_current_canvas, forget_canvas)
  File "/home/runner/work/vispy/vispy/vispy/gloo/__init__.py", line 47, in <module>
    from . import gl  # noqa
  File "/home/runner/work/vispy/vispy/vispy/gloo/gl/__init__.py", line 228, in <module>
    from . import gl2 as default_backend  # noqa
  File "/home/runner/work/vispy/vispy/vispy/gloo/gl/gl2.py", line 26, in <module>
    _lib = ctypes.cdll.LoadLibrary(_lib)
  File "/home/runner/micromamba/envs/vispy-tests/lib/python3.7/ctypes/__init__.py", line 442, in LoadLibrary
    return self._dlltype(name)
  File "/home/runner/micromamba/envs/vispy-tests/lib/python3.7/ctypes/__init__.py", line 364, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: /home/runner/micromamba/envs/vispy-tests/lib/libOSMesa32.so: cannot open shared object file: No such file or directory
make: *** [Makefile:105: osmesa] Error 1
Error: Process completed with exit code 2.
```

larsoner commented:
No idea how to fix this, probably some upstream change in `conda` did something bad or we were using it incorrectly...?

kmuehlbauer commented:
It looks like in recent builds of `mesalib` in conda-forge the lib is named `libOSMesa.so`. But I need to double check.

kmuehlbauer commented:
I've checked a recent build. There is no mention of `libOSMesa32.so` in the most recent build. In older builds there was. So we might fix this in the github workflow file by just using `libOSMesa.so`.

djhoese commented:
@kmuehlbauer Could this be a drop of 32-bit builds by conda-forge (at least for osmesa)?

djhoese commented:
Looks like we can blame @hmaarrfk (:stuck_out_tongue_winking_eye: ) for https://github.com/conda-forge/mesalib-feedstock/pull/33

I don't think it was necessarily anything specific to that PR, but rather that this was probably the first update since conda-forge dropped 32-bit builds. Although I do see some flags in the changed scripts that seem like it is turning off the option for 32-bit compatible builds (not sure if a 32-bit system is required to produce those binaries in this case though).

Either way, not sure why we are looking for the 32-bit .so.

kmuehlbauer commented:
@djhoese it's extremely difficult to trace that on the phone with halfclosed eyes. :grimacing: I'm confident that just using `libOSMesa.so` will do.

djhoese commented:
I've tried switching the lib name in #2245

djhoese commented:
Ugh well it got further but now some of the tests seg fault. :man_facepalming: 

Why does CI do this to me?

djhoese commented:
With version limiting and going back to the `32.so`, tests now pass in that other PR.

kmuehlbauer commented:
@djhoese i've found that issue from back in early 2019: https://github.com/conda-forge/mesalib-feedstock/issues/22

I can't remember why I've requested this. So this might be twofold, first the color-depth issue and second the new mesa version.

@hmaarrfk Do you have any suggestions here?

djhoese commented on vispy/visuals/filters/mesh.py:
```diff
@@ -721,8 +721,10 @@ def _update_data(self):
         self.fshader['faces_only'] = 1 if self._faces_only else 0
-        faces = self._visual.mesh_data.get_faces()
-        n_faces = len(faces)
+        if self._visual.mesh_data.is_empty():
```
Did this need to be changed? Did this work as expected before?

brisvag replied:
This was broken because `get_faces` returns `None` when `Mesh` has no data. It was either this, or checkinf if `faces is None`, so I went for the method just implemented for clarity.

The test I added for `WireframeFilter` does not pass on main, currently.

djhoese replied:
Ah awesome. Thanks.

djhoese commented on vispy/geometry/meshdata.py:
```diff
@@ -654,3 +654,6 @@ def restore(self, state):
                 state[k] = np.array(state[k])
             setattr(self, k, state[k])
+
+    def is_empty(self):
```
Docstring please

```suggestion
    def is_empty(self):
        """Check if any vertices or faces are defined."""
```

djhoese commented on vispy/visuals/filters/tests/test_wireframe_filter.py:
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
```
New files (including the `filters/tests/__init__.py` should copy the license header and copyright.

djhoese reviewed (changes requested):
Small changes requested, otherwise looks good.

djhoese commented:
In case it wasn't clear (@kmuehlbauer), in the other PR I just limited the version of mesalib and went back to 32.so. This mesalib issue is still not fixed.

hmaarrfk commented:
I hope you all don't think this is an issue with the mesalib package. I'm curious what might have caused it.

kmuehlbauer commented:
@hmaarrfk That's what we are trying to investigate. So the old mesalib worked (32.so) but the new mesalib (.so) doesn't work.

djhoese commented:
Note the non-32 .so in the old version also doesn't work. So maybe this is a vispy issue. It seg faulted on CI.

hmaarrfk commented:
hmm it seems that we changed from 32->8 bits without me noticing. https://github.com/conda-forge/mesalib-feedstock/issues/37
